(lib/AST/GenericEnvironment) no generic environment provided for type with type parameters!

crashes on `DEVELOPMENT-SNAPSHOT-2022-09-29-a`:

```swift
protocol P
{
    subscript<Value>(x:Value) -> Int
    {
        get
    }
}
struct S:P
{
    subscript<Value>(x:Int) -> Value
    {
    }
}
```

```
$ swiftc test9.swift 
error: compile command failed due to signal 6 (use -v to see invocation)
test9.swift:12:5: error: missing return in subscript expected to return 'Value'
    }
    ^
test9.swift:8:8: error: type 'S' does not conform to protocol 'P'
struct S:P
       ^
swift-frontend: /home/build-user/swift/lib/AST/GenericEnvironment.cpp:239: static swift::Type swift::GenericEnvironment::mapTypeIntoContext(swift::GenericEnvironment *, swift::Type): Assertion `(env || !type->hasTypeParameter()) && "no generic environment provided for type with type parameters"' failed.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.      Program arguments: /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-29-a/usr/bin/swift-frontend -frontend -c -primary-file test9.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -color-diagnostics -new-driver-path /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-29-a/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-09-29-a/usr/lib/swift -module-name test9 -o /tmp/TemporaryDirectory.ZBcDAU/test9-1.o
1.      Swift version 5.8-dev (LLVM 338153f292479fa, Swift b752d7883c26cea)
2.      Compiling with the current language version
3.      While evaluating request TypeCheckSourceFileRequest(source_file "test9.swift")
4.      While type-checking 'S' (at test9.swift:8:1)
5.      While type-checking protocol conformance to 'P' (at test9.swift:1:1) for type 'S' (declared at [test9.swift:8:1 - line:13:1] RangeText="struct S:P
{
    subscript<Value>(x:Int) -> Value
    {
    }
")
```

--- a/lib/Sema/MiscDiagnostics.cpp
+++ b/lib/Sema/MiscDiagnostics.cpp
@@ -2575,9 +2575,9 @@ static bool fixItOverrideDeclarationTypesImpl(
       });
     }
 
-    auto resultType = subscript->getDeclContext()->mapTypeIntoContext(
-        subscript->getElementInterfaceType());
-    auto baseResultType = baseSubscript->getDeclContext()->mapTypeIntoContext(
+    auto resultType =
+        subscript->mapTypeIntoContext(subscript->getElementInterfaceType());
+    auto baseResultType = baseSubscript->mapTypeIntoContext(
         baseSubscript->getElementInterfaceType());
     fixedAny |= checkType(resultType, ParamDecl::Specifier::Default,
                           baseResultType, ParamDecl::Specifier::Default,

--- a/test/Sema/generic-subscript.swift
+++ b/test/Sema/generic-subscript.swift
@@ -1,0 +1,44 @@
+// RUN: %target-typecheck-verify-swift
+
+
+protocol P
+{
+    subscript<Value>(x:Value) -> Int // expected-note {{protocol requires subscript with type '<Value> (Value) -> Int'; do you want to add a stub?}}
+    {
+        get 
+    } 
+}
+
+struct S:P // expected-error {{type 'S' does not conform to protocol 'P'}}
+{
+    subscript<Value>(x:Int) -> Value // expected-note {{candidate has non-matching type '<Value> (Int) -> Value'}}
+    {
+    } // expected-error {{missing return in subscript expected to return 'Value'}}
+}
+
+
+
+struct S2:P 
+{
+    subscript<Value>(x:Value) -> Int 
+    {
+			return 123;
+    } 
+}
+
+protocol P2
+{
+    subscript(x:Int) -> Int 
+    {
+        get 
+    } 
+}
+
+struct S3:P2
+{
+    subscript(x:Int) -> Int
+    {
+        return x;
+    } 
+}
+

